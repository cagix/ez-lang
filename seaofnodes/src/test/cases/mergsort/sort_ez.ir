START:                            [[   ]]
  10 Start   ____ ____ ____       [[    12   14   17   23    9  162   28  136   59  115  ]]  [  Ctrl, #TOP, int]

L9:                               [[  START      ]]
   9 merge_s ____   10            [[    11   13   16   18   22  213    8  ]]  Ctrl
  11 $mem       9   12            [[     8       ]]  #BOT
  13 a          9   14            [[     8    5  ]]  *[int]
  16 n          9   17            [[     8    5  ]]  int
  22 $rpc       9   23            [[     2       ]]  $[ALL]
  18 b          9   14            [[     8    5  ]]  *[int]
   8 call       9   11   13  213   16   18  [[     7  ]]  Ctrl
 213 #0         9                 [[     8    5  ]]  0

L162:                             [[  START      ]]
 162 main    ____   10            [[   164  212  167  208  171  181  180  179  178  177  176  175  174  173  172  169  193  192  191  190  189  188  187  186  185  184  166  205  204  203  202  201  200  199  198  197  196  163  161  ]]  Ctrl
 164 $mem     162   12            [[   163  171  171  ]]  #BOT
 212 $rpc     162   23            [[   155       ]]  $[ALL]
 181 st8      162  182  183 ____ ____  [[   180  ]]  #3:int
 205 st8      162  206  207 ____ ____  [[   204  ]]  #3:int
 180 st8      162  181  183 ____ ____  [[   179  ]]  #3:int
 204 st8      162  205  207 ____ ____  [[   203  ]]  #3:int
 179 st8      162  180  183 ____ ____  [[   178  ]]  #3:int
 193 st8      162  194  195 ____ ____  [[   192  ]]  #3:int
 203 st8      162  204  207 ____ ____  [[   202  ]]  #3:int
 178 st8      162  179  183 ____ ____  [[   177  ]]  #3:int
 202 st8      162  203  207 ____ ____  [[   201  ]]  #3:int
 192 st8      162  193  195 ____ ____  [[   191  ]]  #3:int
 177 st8      162  178  183 ____ ____  [[   176  ]]  #3:int
 201 st8      162  202  207 ____ ____  [[   200  ]]  #3:int
 191 st8      162  192  195 ____ ____  [[   190  ]]  #3:int
 200 st8      162  201  207 ____ ____  [[   199  ]]  #3:int
 176 st8      162  177  183 ____ ____  [[   175  ]]  #3:int
 190 st8      162  191  195 ____ ____  [[   189  ]]  #3:int
 199 st8      162  200  207 ____ ____  [[   198  ]]  #3:int
 189 st8      162  190  195 ____ ____  [[   188  ]]  #3:int
 175 st8      162  176  183 ____ ____  [[   174  ]]  #3:int
 198 st8      162  199  207 ____ ____  [[   197  ]]  #3:int
 188 st8      162  189  195 ____ ____  [[   187  ]]  #3:int
 174 st8      162  175  183 ____ ____  [[   173  ]]  #3:int
 197 st8      162  198  207 ____ ____  [[   196  ]]  #3:int
 187 st8      162  188  195 ____ ____  [[   186  ]]  #3:int
 173 st8      162  174  183 ____ ____  [[   172  ]]  #3:int
 196 st8      162  197  207 ____ ____  [[   163  ]]  #3:int
 172 st8      162  173  183 ____ ____  [[   169  ]]  #3:int
 186 st8      162  187  195 ____ ____  [[   185  ]]  #3:int
 163 ALLMEM   162  164  165  196  [[   161       ]]  #BOT
 185 st8      162  186  195 ____ ____  [[   184  ]]  #3:int
 169 new_ary  162  167  170  172  [[   168  194  195  ]]  [  Ctrl, *![int], #2:u32, #3:int]
 168 $2       169                 [[   166       ]]  #2:u32
 194 $3       169                 [[   193       ]]  #3:int
 195 [int]    169                 [[   193  192  191  190  189  188  187  186  185  184  161  ]]  *[int]
 161 call     162  163  183  195  208  [[   160  ]]  Ctrl
 184 st8      162  185  195 ____ ____  [[   166  ]]  #3:int
 167 #88      162                 [[   166  169  171  ]]  88
 166 new_ary  162  167  168  184  [[   165  206  207  ]]  [  Ctrl, *![int], #2:u32, #3:int]
 165 $2       166                 [[   163       ]]  #2:u32
 206 $3       166                 [[   205       ]]  #3:int
 207 [int]    166                 [[   205  204  203  202  201  200  199  198  197  196  158  ]]  *[int]
 171 new_ary  162  167  164  164  [[   170  182  183  ]]  [  Ctrl, *![int], #2:u32, #3:int]
 170 $2       171                 [[   169       ]]  #2:u32
 182 $3       171                 [[   181       ]]  #3:int
 183 [int]    171                 [[   181  180  179  178  177  176  175  174  173  172  161  158  ]]  *[int]
 208 #10      162                 [[   161  158  ]]  10

L28:                              [[  START      ]]
  28 split_m ____   10            [[    32   33   44   45   49   54   31   29   27  ]]  Ctrl
  32 end       28   17            [[    31   47   39   36  ]]  int
  33 begin     28   17            [[    31   42   47   36  ]]  int
  44 $mem      28   12            [[    42   52  ]]  #BOT
  45 a         28   14            [[    42   39   36  ]]  *[int]
  54 $rpc      28   23            [[    24       ]]  $[ALL]
  49 b         28   14            [[    42   39   36  ]]  *[int]
  31 sub       28   32   33       [[    29       ]]  int
  29 cmp       28   31            [[    27       ]]  bool
  27 j<=       28   29            [[    26   43  ]]  [  Ctrl, Ctrl]

L136:                             [[  START      ]]
 136 eq      ____   10            [[   142  143  147  150  154  214  135  ]]  Ctrl
 142 $mem     136   12            [[   140  146  152  ]]  #BOT
 143 a        136   14            [[   140       ]]  *[int]
 154 $rpc     136   23            [[   131       ]]  $[ALL]
 150 n        136   17            [[   148       ]]  int
 147 b        136   14            [[   146       ]]  *[int]
 214 #0       136                 [[   153  144  ]]  0

L59:                              [[  START      ]]
  59 merge   ____   10            [[    76   79   82   89   91   96  110  216  217   58  ]]  Ctrl
  76 begin     59   17            [[    75   92  ]]  int
  79 middle    59   17            [[    73   83  ]]  int
  82 end       59   17            [[    80  107  ]]  int
 110 $rpc      59   23            [[    55       ]]  $[ALL]
  96 a         59   14            [[    95   97   86   99  ]]  *[int]
  91 b         59   14            [[    90       ]]  *[int]
  89 $mem      59   12            [[    88  109  ]]  #BOT
 217 #1        59                 [[   105  105  ]]  1
 216 #0        59                 [[   105  105   55  ]]  0

L115:                             [[  START      ]]
 115 copy_ar ____   10            [[   120  122  124  127  129  130  114  ]]  Ctrl
 120 begin    115   17            [[   119       ]]  int
 130 $rpc     115   23            [[   111       ]]  $[ALL]
 129 a        115   14            [[   128       ]]  *[int]
 127 b        115   14            [[   126       ]]  *[int]
 124 $mem     115   12            [[   123  125  ]]  #BOT
 122 end      115   17            [[   117       ]]  int

LOOP135:                          [[  L136      L137       ]]
 135 Loop    ____  136  137       [[   144  148  134  ]]  Ctrl
 144 Phi_i    135  214  145       [[   140  145  146  148  ]]  int
 148 cmp      135  144  150       [[   134       ]]  bool
 134 j<       135  148            [[   133  139  ]]  [  Ctrl, Ctrl]

LOOP58:                           [[  L59       L60        ]]
  58 Loop    ____   59   60       [[    75   83   88   92  107   57  ]]  Ctrl
  75 Phi_i     58   76   77       [[    73   78   77   95   86  ]]  int
  83 Phi_j     58   79   84       [[    80   84   85   97   99  ]]  int
  88 Phi_$3    58   89   90       [[    86   90   95   97   99  109  ]]  #BOT
  92 Phi_k     58   76   93       [[    90   93  107  ]]  int
 107 cmp       58   92   82       [[    57       ]]  bool
  57 j<        58  107            [[    56   72  ]]  [  Ctrl, Ctrl]

LOOP114:                          [[  L115      L116       ]]
 114 Loop    ____  115  116       [[   119  125  117  113  ]]  Ctrl
 119 Phi_k    114  120  121       [[   117  121  126  128  ]]  int
 125 Phi_$3   114  124  126       [[   123  126  128  ]]  #BOT
 117 cmp      114  119  122       [[   113       ]]  bool
 113 j<       114  117            [[   112  116  ]]  [  Ctrl, Ctrl]

L7:                               [[  L9         ]]
   7 CallEnd    8                 [[     6   19  ]]  [  Ctrl, #BOT, int]
  19 $mem       7                 [[     5       ]]  #BOT

L160:                             [[  L162       ]]
 160 CallEnd  161                 [[   159  209  ]]  [  Ctrl, #BOT, int]
 209 $mem     160                 [[   158       ]]  #BOT

L26:                              [[  L28        ]]
  26 True      27                 [[    25       ]]  Ctrl

L43:                              [[  L28        ]]
  43 False     27                 [[    47   48   46   42  ]]  Ctrl
  42 call      43   44   45   33   46   49  [[    41  ]]  Ctrl
  47 add       43   32   33       [[    46       ]]  int
  48 #2        43                 [[    46       ]]  2
  46 div       43   47   48       [[    42   39   36  ]]  int

L6:                               [[  L7         ]]
   6 $ctrl      7                 [[     5       ]]  Ctrl
   5 call       6   19   13  213   16   18  [[     4  ]]  Ctrl

L116:                             [[  LOOP114    ]]
 116 True     113                 [[   128  126  121  114  ]]  Ctrl
 121 inc      116  119            [[   119       ]]  int
 128 ld8      116  125  129  119  [[   126       ]]  int
 126 st8      116  125  127  119  128  [[   125  ]]  #3:Bot

L159:                             [[  L160       ]]
 159 $ctrl    160                 [[   158       ]]  Ctrl
 158 call     159  209  183  207  208  [[   157  ]]  Ctrl

L56:                              [[  LOOP58     ]]
  56 False     57                 [[   109   55  ]]  Ctrl
 109 ALLMEM    56   89 ____   88  [[    55       ]]  #BOT
  55 Return    56  109  216  110  [[     1       ]]  [  Ctrl, #BOT, 0]

L112:                             [[  LOOP114    ]]
 112 False    113                 [[   123  215  111  ]]  Ctrl
 123 ALLMEM   112  124 ____  125  [[   111       ]]  #BOT
 215 #0       112                 [[   111       ]]  0
 111 Return   112  123  215  130  [[     1       ]]  [  Ctrl, #BOT, 0]

L72:                              [[  LOOP58     ]]
  72 True      57                 [[    73   71  ]]  Ctrl
  73 cmp       72   75   79       [[    71       ]]  bool
  71 j<        72   73            [[    70  100  ]]  [  Ctrl, Ctrl]

L139:                             [[  LOOP135    ]]
 139 True     134                 [[   146  140  138  ]]  Ctrl
 146 ld8      139  142  147  144  [[   140       ]]  int
 140 cmp8     139  142  143  144  146  [[   138  ]]  bool
 138 j==      139  140            [[   137  151  ]]  [  Ctrl, Ctrl]

L133:                             [[  LOOP135    ]]
 133 False    134                 [[   106  132  ]]  Ctrl
 106 #1       133                 [[   153       ]]  1

L41:                              [[  L43        ]]
  41 CallEnd   42                 [[    40   50  ]]  [  Ctrl, #BOT, 0]
  50 $mem      41                 [[    39       ]]  #BOT

L4:                               [[  L6         ]]
   4 CallEnd    5                 [[     3   20   21  ]]  [  Ctrl, #BOT, int]
  20 $mem       4                 [[     2       ]]  #BOT
  21 #2         4                 [[     2       ]]  int

L157:                             [[  L159       ]]
 157 CallEnd  158                 [[   156  210  211  ]]  [  Ctrl, #BOT, bool]
 210 $mem     157                 [[   155       ]]  #BOT
 211 #2       157                 [[   155       ]]  bool

L70:                              [[  L72        ]]
  70 True      71                 [[    80   69  ]]  Ctrl
  80 cmp       70   82   83       [[    69       ]]  bool
  69 j<=       70   80            [[    68  101  ]]  [  Ctrl, Ctrl]

L100:                             [[  L72        ]]
 100 False     71                 [[    65       ]]  Ctrl

L137:                             [[  L139       ]]
 137 True     138                 [[   145  135  ]]  Ctrl
 145 inc      137  144            [[   144       ]]  int

L40:                              [[  L41        ]]
  40 $ctrl     41                 [[    39       ]]  Ctrl
  39 call      40   50   45   46   32   49  [[    38  ]]  Ctrl

L151:                             [[  L139       ]]
 151 False    138                 [[   132       ]]  Ctrl

L3:                               [[  L4         ]]
   3 $ctrl      4                 [[     2       ]]  Ctrl
   2 Return     3   20   21   22  [[     1       ]]  [  Ctrl, #BOT, int]

L132:                             [[  L133      L151       ]]
 132 Region  ____  133  151       [[   153  152  131  ]]  Ctrl
 153 Phi_res  132  106  214       [[   131       ]]  bool
 152 ALLMEM   132  142 ____ ____  [[   131       ]]  #BOT
 131 Return   132  152  153  154  [[     1       ]]  [  Ctrl, #BOT, bool]

L156:                             [[  L157       ]]
 156 $ctrl    157                 [[   155       ]]  Ctrl
 155 Return   156  210  211  212  [[     1       ]]  [  Ctrl, #BOT, bool]

L68:                              [[  L70        ]]
  68 False     69                 [[    99   86   67  ]]  Ctrl
  99 ld8       68   88   96   83  [[    86       ]]  int
  86 cmp8      68   88   96   75   99  [[    67  ]]  bool
  67 j<=       68   86            [[    66  102  ]]  [  Ctrl, Ctrl]

L38:                              [[  L40        ]]
  38 CallEnd   39                 [[    37   51  ]]  [  Ctrl, #BOT, 0]
  51 $mem      38                 [[    36       ]]  #BOT

L101:                             [[  L70        ]]
 101 True      69                 [[    65       ]]  Ctrl

L37:                              [[  L38        ]]
  37 $ctrl     38                 [[    36       ]]  Ctrl
  36 call      37   51   49   33   46   32   45  [[    35  ]]  Ctrl

L66:                              [[  L68        ]]
  66 True      67                 [[    65       ]]  Ctrl

L102:                             [[  L68        ]]
 102 False     67                 [[    65       ]]  Ctrl

L65:                              [[  L66       L100      L101      L102       ]]
  65 Region  ____   66  100  101  102  [[   105  103   64  ]]  Ctrl
 105 Phi_con   65  217  216  217  216  [[   103  ]]  bool
 103 test      65  105            [[    64       ]]
  64 j!=       65  103            [[    61   98  ]]  [  Ctrl, Ctrl]

L35:                              [[  L37        ]]
  35 CallEnd   36                 [[    34   53  ]]  [  Ctrl, #BOT, int]
  53 $mem      35                 [[    52       ]]  #BOT

L34:                              [[  L35        ]]
  34 $ctrl     35                 [[    25       ]]  Ctrl

L98:                              [[  L65        ]]
  98 False     64                 [[    97   85   60  ]]  Ctrl
  85 inc       98   83            [[    84       ]]  int
  97 ld8       98   88   96   83  [[    94       ]]  int

L25:                              [[  L26       L34        ]]
  25 Region  ____   26   34       [[    52   15   24  ]]  Ctrl
  52 Phi_$me   25   44   53       [[    24       ]]  #BOT
  15 #0        25                 [[    24       ]]  0
  24 Return    25   52   15   54  [[     1       ]]  [  Ctrl, #BOT, 0]

L61:                              [[  L65        ]]
  61 True      64                 [[    95   78   60  ]]  Ctrl
  78 inc       61   75            [[    77       ]]  int
  95 ld8       61   88   96   75  [[    94       ]]  int

L60:                              [[  L61       L98        ]]
  60 Region  ____   61   98       [[    77   84   94   90   93   58  ]]  Ctrl
  77 Phi_i     60   78   75       [[    75       ]]  int
  84 Phi_j     60   83   85       [[    83       ]]  int
  94 Phi_$3    60   95   97       [[    90       ]]  int
  93 inc       60   92            [[    92       ]]  int
  90 st8       60   88   91   92   94  [[    88  ]]  #3:Bot

L1:                               [[  L112      L132      L156       ]]
   1 Stop       2   24   55  111  131  155  [[   ]]  Bot