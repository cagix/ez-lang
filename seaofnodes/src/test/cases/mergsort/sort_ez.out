---split_merge { *![int] int int *![int] -> int #0}---------------------------
00F0 4883EC38    subi  rsp -= #56
                 a:r9,b:rcx,begin:rdx,end:r8
00F4 4889542428  mov   [rsp+40] = rdx                 // def/loop #1
00F9 4C89442458  mov   [rsp+88] = r8                  // def/loop #0
00FE 48894C2450  mov   [rsp+80] = rcx                 // def/loop #0
0103 4C894C2448  mov   [rsp+72] = r9                  // def/loop #0
0108 4C8B442458  mov   r8 = [rsp+88]                  // use/loop/use #0
010D 4C8B542428  mov   r10 = [rsp+40]                 // use/loop/use #1
0112 4D2BC2      sub   r8 -= r10
0115 4983F801    cmp   r8, #1
0119 7E71        j<=   L65                            // L69
L69:
011B 488B442458  mov   rax = [rsp+88]                 // use/loop/use #0
0120 488B542428  mov   rdx = [rsp+40]                 // use/loop/use #1
0125 4803C2      add   rax += rdx
0128 B902000000  ldi   rcx = #2
012D 489948F7F9  div   rax = rax / rcx                // kill rdx
0132 4889442440  mov   [rsp+64] = rax                 // def/loop #0
0137 488B4C2448  mov   rcx = [rsp+72]                 // use/loop/use #0
013C 4C8B4C2450  mov   r9 = [rsp+80]                  // use/loop/use #0
0141 4C8B442440  mov   r8 = [rsp+64]                  // use/loop/use #0
0146 488B542428  mov   rdx = [rsp+40]                 // use/loop/use #1
014B E8A0FFFFFF  call  split_merge  rcx  rdx  r8  r9
0150 488B4C2448  mov   rcx = [rsp+72]                 // use/loop/use #0
0155 4C8B4C2450  mov   r9 = [rsp+80]                  // use/loop/use #0
015A 4C8B442458  mov   r8 = [rsp+88]                  // use/loop/use #0
015F 488B542440  mov   rdx = [rsp+64]                 // use/loop/use #0
0164 E887FFFFFF  call  split_merge  rcx  rdx  r8  r9
0169 FF7424488F  mov   [rsp+32] = [rsp+72]            // use/loop/use #0
016E 442420
0171 488B4C2450  mov   rcx = [rsp+80]                 // use/loop/use #0
0176 4C8B4C2458  mov   r9 = [rsp+88]                  // use/loop/use #0
017B 4C8B442440  mov   r8 = [rsp+64]                  // use/loop/use #0
0180 488B542428  mov   rdx = [rsp+40]                 // use/loop/use #1
0185 E8A6FEFFFF  call  merge  rcx  rdx  r8  r9  [stk#42]
018A EB02        jmp   L64
L412:
L65:
018C 33C0        xor   rax,rax
L64:
                 arg:rax
018E 4883C438C3  addi  rsp += #56
0193             ret
---{ *![int] int int *![int] -> int #0}---------------------------

---copy_array { *![int] int int *![int] -> int #1}---------------------------
                 begin:rdx,end:r8,b:r9,a:rcx
LOOP157:
                 k:rdx
02F0 493BD0      cmp   rdx, r8
02F3 7D10        j>=   L155                           // L159
L159:
02F5 488B44D108  ld8   rax,[rcx+rdx*8+8]
02FA 498944D108  st8   [r9+rdx*8+8],rax
02FF 4883C201    inc   rdx += #1
0303 EBEB        jmp   LOOP157
L411:
L155:
0305 33C0        xor   rax,rax
0307 C3          ret
---{ *![int] int int *![int] -> int #1}---------------------------

---merge { *![int] int int int *![int] -> int #0}---------------------------
                 begin:rdx,middle:r8,end:r9,b:rcx,a:[rsp+40]
0030 66480F6EC3  mov   xmm0 = rbx                     // def/loop #2
0035 488B5C2428  mov   rbx = [rsp+40]                 // def/empty1 #0
003A 4C8BDA      mov   r11 = rdx                      // use/self/phi #1
003D 4D8BD0      mov   r10 = r8                       // use/self/use #1
LOOP101:
                 i:r11,j:r10,k:rdx
0040 493BD1      cmp   rdx, r9
0043 7D52        j>=   L99                            // L115
L115:
0045 4D3BD8      cmp   r11, r8
0048 7C04        j<    L113                           // L143
L143:
004A 33C0        xor   rax,rax
004C EB24        jmp   L108
L417:
L113:
004E 4D3BCA      cmp   r9, r10
0051 7E1A        j<=   L144                           // L111
L111:
0053 488BC3      mov   rax = rbx                      // use/empty1 #0
0056 4A8B44D008  ld8   rax,[rax+r10*8+8]
005B 4A3944DB08  cmp8  rax, [rbx+r11*8+8]
0060 7E04        j<=   L109                           // L145
L145:
0062 33C0        xor   rax,rax
0064 EB0C        jmp   L108
L416:
L109:
0066 B801000000  ldi   rax = #1
006B EB05        jmp   L108
L415:
L144:
006D B801000000  ldi   rax = #1
L108:
                 cond:rax
0072 4883F800    test  rax
0076 750B        j!=   L104                           // L141
L141:
0078 4A8B44D308  ld8   rax,[rbx+r10*8+8]
007D 4983C201    inc   r10 += #1
0081 EB09        jmp   L103
L414:
L104:
0083 4A8B44DB08  ld8   rax,[rbx+r11*8+8]
0088 4983C301    inc   r11 += #1
L103:
                 i:r11,j:r10,$3:rax
008C 488944D108  st8   [rcx+rdx*8+8],rax
0091 4883C201    inc   rdx += #1
0095 EBA9        jmp   LOOP101
L413:
L99:
0097 66480F7EC3  mov   rbx = xmm0                     // use/loop/use #2
009C 33C0        xor   rax,rax
009E C3          ret
---{ *![int] int int int *![int] -> int #0}---------------------------


---eq { *![int] *![int] int -> int #1}---------------------------
                 a:rcx,b:rdx,n:r8
0000 33C0        xor   rax,rax
LOOP178:
                 i:rax
0002 493BC0      cmp   rax, r8
0005 7D16        j>=   L176                           // L186
L186:
0007 4C8B4CC208  ld8   r9,[rdx+rax*8+8]
000C 4C394CC108  cmp8  r9, [rcx+rax*8+8]
0011 7506        j!=   L198                           // L184
L184:
0013 4883C001    inc   rax += #1
0017 EBE9        jmp   LOOP178
L419:
L198:
0019 33C0        xor   rax,rax
001B EB05        jmp   L175
L418:
L176:
001D B801000000  ldi   rax = #1
L175:
                 result:rax
0022 C3          ret
---{ *![int] *![int] int -> int #1}---------------------------

---main { -> int #0}---------------------------
01A0 4883EC28    subi  rsp -= #40
01A4 48895C2420  mov   [rsp+32] = rbx                 // def/loop #2
01A9 BA58000000  ldi   rdx = #88
01AE B901000000  alloc ldi   rcx = #1
01B3 E800000000  call  #calloc
01B8 4889442418  mov   [rsp+24] = rax                 // def/loop #0
01BD 488B542418  mov   rdx = [rsp+24]                 // use/loop/use #0
01C2 48C742080A  st8   [rdx+8],#10
01C7 000000
01CA 488B4C2418  mov   rcx = [rsp+24]                 // use/loop/use #0
01CF 48C7411009  st8   [rcx+16],#9
01D4 000000
01D7 488B442418  mov   rax = [rsp+24]                 // use/loop/use #0
01DC 48C7401808  st8   [rax+24],#8
01E1 000000
01E4 4C8B5C2418  mov   r11 = [rsp+24]                 // use/loop/use #0
01E9 49C7432007  st8   [r11+32],#7
01EE 000000
01F1 4C8B542418  mov   r10 = [rsp+24]                 // use/loop/use #0
01F6 49C7422806  st8   [r10+40],#6
01FB 000000
01FE 4C8B4C2418  mov   r9 = [rsp+24]                  // use/loop/use #0
0203 49C7413005  st8   [r9+48],#5
0208 000000
020B 4C8B442418  mov   r8 = [rsp+24]                  // use/loop/use #0
0210 49C7403804  st8   [r8+56],#4
0215 000000
0218 488B542418  mov   rdx = [rsp+24]                 // use/loop/use #0
021D 48C7424003  st8   [rdx+64],#3
0222 000000
0225 488B4C2418  mov   rcx = [rsp+24]                 // use/loop/use #0
022A 48C7414802  st8   [rcx+72],#2
022F 000000
0232 488B442418  mov   rax = [rsp+24]                 // use/loop/use #0
0237 48C7405001  st8   [rax+80],#1
023C 000000
023F BA58000000  ldi   rdx = #88
0244 B901000000  alloc ldi   rcx = #1
0249 E800000000  call  #calloc
024E 4889442410  mov   [rsp+16] = rax                 // def/empty1 #1
0253 BA58000000  ldi   rdx = #88
0258 B901000000  alloc ldi   rcx = #1
025D E800000000  call  #calloc
0262 488BD8      mov   rbx = rax                      // def/empty1 #0
0265 48C7430801  st8   [rbx+8],#1
026A 000000
026D 48C7431002  st8   [rbx+16],#2
0272 000000
0275 48C7431803  st8   [rbx+24],#3
027A 000000
027D 48C7432004  st8   [rbx+32],#4
0282 000000
0285 48C7432805  st8   [rbx+40],#5
028A 000000
028D 48C7433006  st8   [rbx+48],#6
0292 000000
0295 48C7433807  st8   [rbx+56],#7
029A 000000
029D 48C7434008  st8   [rbx+64],#8
02A2 000000
02A5 48C7434809  st8   [rbx+72],#9
02AA 000000
02AD 48C743500A  st8   [rbx+80],#10
02B2 000000
02B5 488B4C2418  mov   rcx = [rsp+24]                 // use/loop/use #0
02BA 488B542410  mov   rdx = [rsp+16]                 // use/empty1 #0
02BF 41B80A0000  ldi   r8 = #10
02C4 00
02C5 E8D6FDFFFF  call  merge_sort  rcx  rdx  r8
02CA 488B4C2418  mov   rcx = [rsp+24]                 // use/loop/use #0
02CF 488BD3      mov   rdx = rbx                      // use/empty1 #0
02D2 41B80A0000  ldi   r8 = #10
02D7 00
02D8 E823FDFFFF  call  eq  rcx  rdx  r8
02DD 488B5C2420  mov   rbx = [rsp+32]                 // use/loop/use #2
02E2 4883C428C3  addi  rsp += #40
02E7             ret
---{ -> int #0}---------------------------

---merge_sort { *![int] *![int] int -> int #0}---------------------------
00A0 4883EC28    subi  rsp -= #40
                 a:rcx,n:r8,b:rdx
00A4 4C89442440  mov   [rsp+64] = r8                  // def/loop #1
00A9 48894C2438  mov   [rsp+56] = rcx                 // def/loop #1
00AE 4889542430  mov   [rsp+48] = rdx                 // def/loop #0
00B3 4C8B4C2430  mov   r9 = [rsp+48]                  // use/loop/use #0
00B8 488B4C2438  mov   rcx = [rsp+56]                 // use/loop/use #1
00BD 4C8B442440  mov   r8 = [rsp+64]                  // use/loop/use #1
00C2 33D2        xor   rdx,rdx
00C4 E827020000  call  copy_array  rcx  rdx  r8  r9
00C9 4C8B4C2430  mov   r9 = [rsp+48]                  // use/loop/use #0
00CE 488B4C2438  mov   rcx = [rsp+56]                 // use/loop/use #1
00D3 4C8B442440  mov   r8 = [rsp+64]                  // use/loop/use #1
00D8 33D2        xor   rdx,rdx
00DA E811000000  call  split_merge  rcx  rdx  r8  r9
00DF 4883C428C3  addi  rsp += #40
00E4             ret
---{ *![int] *![int] int -> int #0}---------------------------


